x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -463,27 +478,40 @@ export class ExecutionHandler<
     };
   }
 
-  private generateExecutables(
+  private async generateExecutables(
     alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>,
-    summaryActions: ThrottledActions
-  ) {
+    throttledSummaryActions: ThrottledActions
+  ): Promise<Array<Executable<State, Context, ActionGroupIds, RecoveryActionGroupId>>> {
     const executables = [];
 
     for (const action of this.rule.actions) {
+      const alertsArray = Object.entries(alerts);
+      const summarizedAlerts = await this.getSummarizedAlerts({
+        action,
+        spaceId: this.taskInstance.params.spaceId,
+        ruleId: this.taskInstance.params.alertId,
+        throttledSummaryActions,
+      });
+
+      if (summarizedAlerts !== null) {
+        this.logNumberOfFilteredAlerts({
+          numberOfAlerts: alertsArray.length,
+          numberOfSummarizedAlerts: summarizedAlerts.all.count,
+          action,
+        });
+      }
+
       if (isSummaryAction(action)) {
-        if (
-          this.canFetchSummarizedAlerts(action) &&
-          !isSummaryActionThrottled({
-            action,
-            summaryActions,
-            logger: this.logger,
-          })
-        ) {
-          executables.push({ action });
+        if (summarizedAlerts !== null) {
+          executables.push({ action, summarizedAlerts });
         }
         continue;
       }
-      for (const [alertId, alert] of Object.entries(alerts)) {
+
+      for (const [alertId, alert] of alertsArray) {
+        if (alert.isFilteredOut(summarizedAlerts)) {
```

Comment:
Does this check mean if the action is not a summary action, we are still applying the alerts filter? So any action attached to a rule that summarizes actions will run this additional query?

I wonder if it would be more performant to include the alert ids in the query instead of querying for everything and checking afterward?

Reply:
> Does this check mean if the action is not a summary action, we are still applying the alerts filter? 
yes, this is to filter the non-summary alerts. And will change once we switch to alert-as-data for all the alerts.

> So any action attached to a rule that summarizes actions will run this additional query?
No, only the ones that have alertsFilter. `getSummarizedAlerts` returns `null` for the alerts that are not summary and has no alertsFilter. And isFilteredOut checks if summaryAlerts is null. just didn't want to add multiple if-checks here and moved all of them in to either `getSummarizedAlerts` or `isFilteredOut`

> I wonder if it would be more performant to include the alert ids in the query instead of querying for everything and checking afterward?
I am not sure if i get this correctly, you mean passing alertIds to `getSummarizedAlerts` ?
If so, summarizedAlerts are already filtered, we use them to filter the ones in memory as well.

Reply:
I realised that figuring out this is a bit difficult, therefore moved the getting summarized alerts decision logic from `getSummarizedAlerts` and wrapped it with `shouldGetSummarizedAlerts`